File: ruby/clock/clock.rb
class Clock
  def initialize(hour: 0, minute: 0)
    @minute = (minute + hour * 60) % MINUTES_IN_DAY
  end

  def to_s
     TIME_FORMAT % to_hour_min
  end

  def +(clock)
    Clock.new(minute: minute + clock.minute)
  end

  def -(clock)
    Clock.new(minute: minute - clock.minute)
  end

  def ==(clock)
    minute == clock.minute
  end

  protected
  attr_reader :minute

  def to_hour_min
    minute.divmod 60
  end

  private
  MINUTES_IN_DAY = 24 * 60
  TIME_FORMAT = "%02d:%02d"
end

File: haskell/clock/src/Clock.hs
module Clock (addDelta, fromHourMin, toString) where

data Clock = Clock Int Int
  deriving Eq

-- Clock 을 생성할 때 아래 제약을 만족하게 변환
-- 0 <= min <= 60 && 0 <= hour <= 24
fromHourMin :: Int -> Int -> Clock
fromHourMin hour min = Clock (mod (hour + q) 24) r
  where (q, r) = divMod min 60

format :: Int -> String
format num
  | length str == 1 = "0" ++ str
  | otherwise = str
  where str = show num

toString :: Clock -> String
toString (Clock hour min) =
  format hour ++ ":" ++ format min

addDelta :: Int -> Int -> Clock -> Clock
addDelta hour min =
    add (Clock hour min)

add :: Clock -> Clock -> Clock
add (Clock hour1 min1) (Clock hour2 min2) =
  fromHourMin (hour1 + hour2) (min1 + min2)
